sdk/messaging/azservicebus/receiver.go

```diff
@@ -357,141 +372,69 @@ func (r *Receiver) receiveMessagesImpl(ctx context.Context, maxMessages int, opt
-				log.Writef(EventReceiver, "Fatal error when receiving: %#v", lastErr)
-				return nil, lastErr
-			}
+	// request just the right amount of credits, taking into account the credits
+	// that are already active on the link from prior ReceiveMessages() calls that
+	// might have exited before all credits were used up.
+	currentReceiverCredits := int64(linksWithID.Receiver.Credits())
+	creditsToIssue := int64(maxMessages) - currentReceiverCredits
+	log.Writef(EventReceiver, "Asking for %d credits", maxMessages)
 
-			// there's no material difference here, so let them fail on the next call to ReceiveMessages() instead.
-			log.Writef(EventReceiver, "Non-fatal error occurred and no messages were received. Will recover on next call: %v", lastErr)
-			return nil, nil
-		}
-	}
+	if creditsToIssue > 0 {
+		log.Writef(EventReceiver, "Only need to issue %d additional credits", creditsToIssue)
 
-	if err := fetchMessages(ctx, linksWithID.Receiver, maxMessages, r.defaultTimeAfterFirstMsg, &all); err != nil {
-		// if the user's cancelled the fetch we'll fall through and let the drain happen.
-		if !internal.IsCancelError(err) {
-			// If the user didn't cancel then we had an actual failure that's going to require a
-			// link/connection reset. Close links and return whatever we can.
-			cleanup(err)
-			return flushAndReturn(err)
+		if err := linksWithID.Receiver.IssueCredit(uint32(creditsToIssue)); err != nil {
+			return nil, err
 		}
-
-		// user cancelled (or we hit our "afte first message" timer).
-		// Fall through so we drain the link since it's still active.
-		log.Writef(EventReceiver, "Context has been cancelled")
+	} else {
+		log.Writef(EventReceiver, "No additional credits needed, still have %d credits active", currentReceiverCredits)
 	}
 
-	if len(all) < maxMessages { // drain if there are excess credits
-		log.Writef(EventReceiver, "Draining to clear %d credits", maxMessages-len(all))
-
-		drainCtx, drainCtxcancel := context.WithTimeout(context.Background(), r.defaultDrainTimeout)
-		defer drainCtxcancel()
-
-		// Start the drain asynchronously. Note that we ignore the user's context at this point
-		// since draining makes sure we don't get messages when nobody is receiving.
-		if err := linksWithID.Receiver.DrainCredit(drainCtx); err != nil {
-			// Cancelling a DrainCredit means we're in an indeterminate state
-			// so we treat that as a "must recover" error.
-			cleanup(err)
-
-			if internal.IsCancelError(err) {
-				log.Writef(EventReceiver, "Drain didn't complete in %s, link was recreated", r.defaultDrainTimeout)
-
-				// this cancellation is entirely on us - we cleaned up as best we can
-				// but the user isn't expected to know about this detail so we don't
-				// return the cancellation error.
-				err = nil
-			}
-
-			return flushAndReturn(err)
-		}
+	result := r.fetchMessages(ctx, linksWithID.Receiver, maxMessages, r.defaultTimeAfterFirstMsg)
 
-		return flushAndReturn(ctx.Err())
+	if result.Messages != nil {
```

Comment:
You don't want it to log if the message count is zero?

Reply:
That's a good point, will fix.